# Multiple Linear Regression in Statsmodels - Lab

## Introduction
In this lab, you'll practice fitting a multiple linear regression model on the Boston Housing dataset!

## Objectives
You will be able to:
* Determine if it is necessary to perform normalization/standardization for a specific model or set of data
* Use standardization/normalization on features of a dataset
* Identify if it is necessary to perform log transformations on a set of features
* Perform log transformations on different features of a dataset
* Use statsmodels to fit a multiple linear regression model
* Evaluate a linear regression model by using statistical performance metrics pertaining to overall model and specific parameters


## The Boston Housing Data

We pre-processed the Boston Housing data again. This time, however, we did things slightly different:
- We dropped `'ZN'` and `'NOX'` completely 
- We categorized `'RAD'` in 3 bins and `'TAX'` in 4 bins
- We transformed `'RAD'` and `'TAX'` to dummy variables and dropped the first variable to eliminate multicollinearity
- We used min-max-scaling on `'B'`, `'CRIM'`, and `'DIS'` (and log transformed all of them first, except `'B'`)
- We used standardization on `'AGE'`, `'INDUS'`, `'LSTAT'`, and `'PTRATIO'` (and log transformed all of them first, except for `'AGE'`) 

In [4]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_boston
boston = load_boston()

boston_features = pd.DataFrame(boston.data, columns = boston.feature_names)
boston_features = boston_features.drop(['NOX', 'ZN'],axis=1)

# First, create bins for based on the values observed. 3 values will result in 2 bins
bins = [0, 6, 24]
bins_rad = pd.cut(boston_features['RAD'], bins)
bins_rad = bins_rad.cat.as_unordered()

# First, create bins for based on the values observed. 4 values will result in 3 bins
bins = [0, 270, 360, 712]
bins_tax = pd.cut(boston_features['TAX'], bins)
bins_tax = bins_tax.cat.as_unordered()

tax_dummy = pd.get_dummies(bins_tax, prefix='TAX', drop_first=True)
rad_dummy = pd.get_dummies(bins_rad, prefix='RAD', drop_first=True)
boston_features = boston_features.drop(['RAD','TAX'], axis=1)
boston_features = pd.concat([boston_features, rad_dummy, tax_dummy], axis=1)

In [5]:
age = boston_features['AGE']
b = boston_features['B']
logcrim = np.log(boston_features['CRIM'])
logdis = np.log(boston_features['DIS'])
logindus = np.log(boston_features['INDUS'])
loglstat = np.log(boston_features['LSTAT'])
logptratio = np.log(boston_features['PTRATIO'])

# Min-Max scaling
boston_features['B'] = (b-min(b))/(max(b)-min(b))
boston_features['CRIM'] = (logcrim-min(logcrim))/(max(logcrim)-min(logcrim))
boston_features['DIS'] = (logdis-min(logdis))/(max(logdis)-min(logdis))

# Standardization
boston_features['AGE'] = (age-np.mean(age))/np.sqrt(np.var(age))
boston_features['INDUS'] = (logindus-np.mean(logindus))/np.sqrt(np.var(logindus))
boston_features['LSTAT'] = (loglstat-np.mean(loglstat))/np.sqrt(np.var(loglstat))
boston_features['PTRATIO'] = (logptratio-np.mean(logptratio))/(np.sqrt(np.var(logptratio)))

In [6]:
boston_features.head()

CRIM     INDUS  CHAS     RM       AGE       DIS   PTRATIO         B  \
0  0.000000 -1.704344   0.0  6.575 -0.120013  0.542096 -1.443977  1.000000   
1  0.153211 -0.263239   0.0  6.421  0.367166  0.623954 -0.230278  1.000000   
2  0.153134 -0.263239   0.0  7.185 -0.265812  0.623954 -0.230278  0.989737   
3  0.171005 -1.778965   0.0  6.998 -0.809889  0.707895  0.165279  0.994276   
4  0.250315 -1.778965   0.0  7.147 -0.511180  0.707895  0.165279  1.000000   

      LSTAT  RAD_(6, 24]  TAX_(270, 360]  TAX_(360, 712]  
0 -1.275260            0               1               0  
1 -0.263711            0               0               0  
2 -1.627858            0               0               0  
3 -2.153192            0               0               0  
4 -1.162114            0               0               0

In [8]:
boston_features.columns

Index(['CRIM', 'INDUS', 'CHAS', 'RM', 'AGE', 'DIS', 'PTRATIO', 'B', 'LSTAT',
       'RAD_(6, 24]', 'TAX_(270, 360]', 'TAX_(360, 712]'],
      dtype='object')

In [35]:
new_data = boston_features.rename(columns={"RAD_(6, 24]": "RAD1", "TAX_(270, 360]": "TAX1","TAX_(360, 712]":"TAX2"})

## Run a linear model in statsmodels

In [23]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

In [38]:
# Your code here

# formula = "LSTAT ~ CRIM+INDUS+CHAS+RM+AGE+DIS+PTRATIO+B+RAD1+TAX1+TAX2"
# model = ols(formula= formula, data=new_data).fit()
outcome = 'LSTAT'
predictors = new_data.drop('LSTAT', axis=1)
pred_sum = "+".join(predictors.columns)
formula1 = outcome + "~" + pred_sum


In [41]:
model = ols(formula= formula1, data=new_data).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  LSTAT   R-squared:                       0.698
Model:                            OLS   Adj. R-squared:                  0.691
Method:                 Least Squares   F-statistic:                     103.7
Date:                Thu, 28 May 2020   Prob (F-statistic):          1.07e-120
Time:                        10:55:53   Log-Likelihood:                -415.29
No. Observations:                 506   AIC:                             854.6
Df Residuals:                     494   BIC:                             905.3
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      4.1990      0.359     11.699      0.000       3.494       4.904
CRIM           0.5981      0.268      2.235      0.026       0.072       1.124
INDUS          0.1027      0.046      2.238      0.026       0.013       0.193
CHAS          -0.2183      0.101     -2.166      0.031      -0.416      -0.020
RM            -0.6819      0.042    -16.296      0.000      -0.764      -0.600
AGE            0.4021      0.041      9.831      0.000       0.322       0.483
DIS            0.5471      0.235      2.330      0.020       0.086       1.008
PTRATIO        0.0517      0.031      1.688      0.092      -0.008       0.112
B             -0.4466      0.124     -3.609      0.000      -0.690      -0.203
RAD1          -0.0123      0.086     -0.144      0.886      -0.180       0.156
TAX1          -0.0100      0.076     -0.131      0.896      -0.160       0.140
TAX2          -0.0079      0.090     -0.088      0.930      -0.184       0.168
==============================================================================
Omnibus:                      163.499   Durbin-Watson:                   1.092
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              799.337
Skew:                          -1.343   Prob(JB):                    2.67e-174
Kurtosis:                       8.541   Cond. No.                         110.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Run the same model in scikit-learn

In [42]:
# Your code here - Check that the coefficients and intercept are the same as those from Statsmodels
from sklearn.linear_model import LinearRegression
y = data_ols['LSTAT']
linreg = LinearRegression()
linreg.fit(predictors, y)

NameError: name 'data_ols' is not defined

## Interpret the coefficients for PTRATIO, PTRATIO, LSTAT

- CRIM: per capita crime rate by town
- INDUS: proportion of non-retail business acres per town
- CHAS: Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
- RM: average number of rooms per dwelling
- AGE: proportion of owner-occupied units built prior to 1940
- DIS: weighted distances to five Boston employment centres
- RAD: index of accessibility to radial highways
- TAX: full-value property-tax rate per $10,000
- PTRATIO: pupil-teacher ratio by town
- B: 1000(Bk - 0.63)^2 where Bk is the proportion of African American individuals by town
- LSTAT: % lower status of the population

## Predict the house price given the following characteristics (before manipulation!!)

Make sure to transform your variables as needed!

- CRIM: 0.15
- INDUS: 6.07
- CHAS: 1        
- RM:  6.1
- AGE: 33.2
- DIS: 7.6
- PTRATIO: 17
- B: 383
- LSTAT: 10.87
- RAD: 8
- TAX: 284

## Summary
Congratulations! You pre-processed the Boston Housing data using scaling and standardization. You also fitted your first multiple linear regression model on the Boston Housing data using statsmodels and scikit-learn!